Applications/Spire/Source/Ui/RealSpinBox.cpp

```diff
@@ -0,0 +1,441 @@
+#include "Spire/Ui/RealSpinBox.hpp"
+#include <QApplication>
+#include <QKeyEvent>
+#include <QLineEdit>
+#include <QStyleOption>
+#include "Spire/Spire/Dimensions.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  const auto SHIFT_STEPS = 10;
+  const auto MOUSE_REPEAT_DELAY_MS = 20;
+  const auto DOWN_ARROW_ICON = ":/Icons/arrow-down.svg";
+  const auto DOWN_ARROW_HOVER_ICON = ":/Icons/arrow-down-hover.svg";
+  const auto DOWN_ARROW_DISABLED_ICON = ":/Icons/arrow-down-disabled.svg";
+  const auto UP_ARROW_ICON = ":/Icons/arrow-up.svg";
+  const auto UP_ARROW_HOVER_ICON = ":/Icons/arrow-up-hover.svg";
+  const auto UP_ARROW_DISABLED_ICON = ":/Icons/arrow-up-disabled.svg";
+
+  auto ARROW_WIDTH() {
+    static auto width = scale_width(8);
+    return width;
+  }
+
+  auto clamped_real(RealSpinBox::Real value, RealSpinBox::Real min,
```

Comment:
Can you use ```std::clamp```?

Reply:
I could use std::clamp but I would still wrap it in clamped_real because it would require a comparison function.

Reply:
How would you use it? ```Real``` doesn't have an ```operator <```?

Reply:
Boost implemented increment/decrement, equality, and the arithmetic assignment operators, but I don't see the other comparison operators. The cpp_dec_float class has the compare function with the comment:
`   // Return +1 for *this > v`
`   // 0 for *this = v`
`   // -1 for *this < v`

So clamped_real(...) would wrap a call to std::clamp that takes a comparison function with a call to cpp_dec_float::compare(...).